packages/server/src/main/java/software/uncharted/terarium/hmiserver/service/data/SimulationService.java

```diff
@@ -81,6 +88,55 @@ public SimulationUpdate appendUpdateToSimulation(
 		return created;
 	}
 
+	public Dataset createDatasetFromSimulation(
+		final UUID simId,
+		final String datasetName,
+		final UUID projectId,
+		final boolean addToProject,
+		final Schema.Permission permission
+	) {
+		try {
+			final Optional<Simulation> sim = this.getAsset(simId, permission);
+			if (sim.isEmpty()) {
+				throw new IllegalArgumentException("Simulation not found");
+			}
+
+			final Dataset dataset = datasetService.createAsset(new Dataset(), projectId, permission);
+			dataset.setName(datasetName);
+			dataset.setDescription(sim.get().getDescription());
+			dataset.setMetadata(objectMapper.convertValue(Map.of("simulationId", simId.toString()), JsonNode.class));
+			dataset.setFileNames(sim.get().getResultFiles());
+			dataset.setDataSourceDate(sim.get().getCompletedTime());
+			dataset.setColumns(new ArrayList<>());
+
+			// Attach the user to the dataset
+			if (sim.get().getUserId() != null) {
+				dataset.setUserId(sim.get().getUserId());
+			}
+
+			// Duplicate the simulation results to a new dataset
+			this.copySimulationResultToDataset(sim.get(), dataset);
+			datasetService.updateAsset(dataset, projectId, permission);
+
+			if (!addToProject) {
+				return dataset;
+			}
+
+			// Add the dataset to the project
+			final Optional<Project> project = projectService.getProject(projectId);
+			if (project.isPresent()) {
```

Comment:
The above code isn't too expensive, but, it _will_ create a dataset in the database. If this project check fails then this dataset would be orphaned. If that's not expected then we may want to check for the presence of the project and simulation right at the start to know that we can proceed

Reply:
makes sense - Ill make those changes

Reply:
I've added the project check at the start of the function